project/JavalibIRCleaner.scala

```diff
@@ -0,0 +1,435 @@
+        if (path.getFileName().toString().endsWith(".sjsir"))
+          builder += new IRFile(path, readIRFile(path))
+        super.visitFile(path, attrs)
+      }
+    }
+
+    Files.walkFileTree(path, ju.EnumSet.of(FileVisitOption.FOLLOW_LINKS),
+        Int.MaxValue, dirVisitor)
+    builder.result()
+  }
+
+  private def readIRFile(path: Path): ClassDef = {
+    import java.nio.ByteBuffer
+    import java.nio.channels.FileChannel
+
+    val channel = FileChannel.open(path)
+    try {
+      val fileSize = channel.size()
+      if (fileSize > Int.MaxValue.toLong)
+        throw new IOException(s"IR file too large: $path")
+      val buffer = ByteBuffer.allocate(fileSize.toInt)
+      channel.read(buffer)
+      buffer.flip()
+      Serializers.deserialize(buffer)
+    } finally {
+      channel.close()
+    }
+  }
+
+  private def writeIRFile(path: Path, tree: ClassDef): Unit = {
+    Files.createDirectories(path.getParent())
+    val outputStream =
+      new BufferedOutputStream(new FileOutputStream(path.toFile()))
+    try {
+      Serializers.serialize(outputStream, tree)
+    } finally {
+      outputStream.close()
+    }
+  }
+
+  private def analyzeJSTypes(irFiles: Seq[IRFile]): Map[String, ClassDef] = {
+    (for {
+      irFile <- irFiles
+      if irFile.tree.kind.isJSType
+    } yield {
+      val tree = irFile.tree
+      tree.name.encodedName -> tree
+    }).toMap
+  }
+
+  private def cleanTree(tree: ClassDef, jsTypes: Map[String, ClassDef],
+      errorManager: ErrorManager): ClassDef = {
+    new ClassDefCleaner(tree.name.name, jsTypes, errorManager)
+      .cleanClassDef(tree)
+  }
+
+  private final class ClassDefCleaner(enclosingClassName: String,
+      jsTypes: Map[String, ClassDef], errorManager: ErrorManager)
+      extends Transformers.ClassTransformer {
+
+    def cleanClassDef(tree: ClassDef): ClassDef = {
+      import tree._
+
+      val preprocessedTree = {
+        var changed = false
+
+        // Preprocess the super interface list
+        val newInterfaces = transformInterfaceList(interfaces)
+        if (newInterfaces ne interfaces)
+          changed = true
+
+        /* Remove the `private def writeReplace__O` generated by scalac 2.13+
+         * in the companion of serializable classes.
+         */
+        val newMemberDefs = memberDefs.filter { m =>
+          if (m.name.encodedName == "writeReplace__O") {
+            changed = true
+            false
+          } else {
+            true
+          }
+        }
+
+        // Check that no two methods have been erased to the same name
+        val seenMethodNames = mutable.Set.empty[(MemberNamespace, String)]
+        for (m <- newMemberDefs) {
+          if (m.isInstanceOf[MethodDef]) {
+            if (!seenMethodNames.add((m.flags.namespace, m.name.encodedName))) {
+              reportError(
+                  s"duplicate method name ${m.name.encodedName} after erasure")(
+                  m.pos)
+            }
+          }
+        }
```

Comment:
Hum, is it me or is this pre-transform of the methods? Or am I reading this wrong?

Reply:
It iterates over `newMemberDefs` doesn't it?

Reply:
Er ... Yes, you're right -_-'